src/lib/plugin_apis/crypto.api

```diff
@@ -1111,6 +1112,193 @@ gboolean bd_crypto_luks_set_uuid (const gchar *device, const gchar *uuid, GError
+ *                                  i.e. whether a device is active in the time of reencryption or not
+ * @pbkdf: (nullable):          PBDKF function parameters for a new keyslot
+ */
+BDCryptoLUKSReencryptParams* bd_crypto_luks_reencrypt_params_new (guint32 key_size, gchar *cipher, gchar *cipher_mode, gchar *resilience, gchar *hash, guint64 max_hotzone_size, guint32 sector_size, gboolean new_volume_key, gboolean offline, BDCryptoLUKSPBKDF *pbkdf) {
+    BDCryptoLUKSReencryptParams *ret = g_new0 (BDCryptoLUKSReencryptParams, 1);
+    ret->key_size = key_size;
+    ret->cipher = g_strdup (cipher);
+    ret->cipher_mode = g_strdup (cipher_mode);
+    ret->resilience = g_strdup (resilience);
+    ret->hash = g_strdup (hash);
+    ret->max_hotzone_size = max_hotzone_size;
+    ret->sector_size = sector_size;
+    ret->new_volume_key = new_volume_key;
+    ret->offline = offline;
+    ret->pbkdf = bd_crypto_luks_pbkdf_copy(pbkdf);
+
+    return ret;
+}
+
+GType bd_crypto_luks_reencrypt_params_get_type () {
+    static GType type = 0;
+
+    if (G_UNLIKELY(type == 0)) {
+        type = g_boxed_type_register_static("BDCryptoLUKSReencryptParams",
+                                            (GBoxedCopyFunc) bd_crypto_luks_reencrypt_params_copy,
+                                            (GBoxedFreeFunc) bd_crypto_luks_reencrypt_params_free);
+    }
+
+    return type;
+}
+
+/**
+ * BDCryptoLUKSReencryptProgFunc:
+ * @size    size of the device being reencrypted
+ * @offset  current offset
+ *
+ * A callback function called during reencryption to report progress. Also used to possibly stop reencryption.
+ *
+ * Returns: 0, if the reencryption should continue.
+ *          A non-zero value to stop the reencryption
+ */
+typedef int (*BDCryptoLUKSReencryptProgFunc) (guint64 size, guint64 offset);
+
+typedef enum {
+    BD_CRYPTO_LUKS_REENCRYPT_NONE = 0,
+    BD_CRYPTO_LUKS_REENCRYPT_CLEAN,
+    BD_CRYPTO_LUKS_REENCRYPT_CRASH,
+    BD_CRYPTO_LUKS_REENCRYPT_INVALID
+} BDCryptoLUKSReencryptStatus;
+
+typedef enum {
+    BD_CRYPTO_LUKS_REENCRYPT = 0,
+    BD_CRYPTO_LUKS_ENCRYPT,
+    BD_CRYPTO_LUKS_DECRYPT,
+} BDCryptoLUKSReencryptMode;
+
+/**
+ * bd_crypto_luks_reencrypt:
+ * @device:     device to reencrypt. Either an active device name for online reencryption, or a block device for offline reencryption.
+ *              Must match the @params's "offline" parameter
+ * @params:     reencryption parameters
+ * @context:    key slot context to unlock @device. The newly created keyslot will use the same context
+ * @prog_func: (scope call) (nullable): progress function. Also used to possibly stop reencryption
+ * @error: (out) (optional): place to store error (if any)
+ *
+ * Reencrypts @device. This could mean a change of cipher, cipher mode, or volume key, based on @params
+ *
+ * Returns: true,  if the reencryption was successful or gracefully stopped with @prog_func.
+ *          false, if an error occurred.
```

Comment:
As this is a long-running operation, I (as an user) would like to know whether this is a blocking call or not. I.e. whether this is a synchronous operation or just a trigger to start an operation in the background. If this is a blocking call, would be good to warn user that this may take hours or days to complete.